scripts/multiplayer.gd

```diff
@@ -0,0 +1,179 @@
+	ip_adress.visible = false
+
+func _on_Connect_pressed():
+	if ip_adress.get_node("LineEdit").text == "":
+		return
+	ip_adress.visible = false
+	SERVER_IP = ip_adress.get_node("LineEdit").text
+	start_multiplayer("client")
+
+func _on_BackButton_pressed():
+	get_tree().network_peer = null
+	get_tree().change_scene("res://menus/playmenu.tscn")
+
+func _process(delta):
+	if game_running:
+		#Update player position
+		var character = game.get_node("romp_online/player/Player/KinematicBody")
+		rpc_unreliable("position_update", character.global_transform.origin, character.rotation)
+		if mp_type != "server":
+			return
+		#The server tells the clients the ball positions
+		var targets = game.get_node("romp_online/Targets")
+		var balls_pos_array = []
+		for i in targets.get_children():
+			balls_pos_array.append(i.global_transform.origin)
+		rpc_unreliable("targets_update", balls_pos_array)
+
+remote func targets_update(pos_array):
+	var targets = game.get_node("romp_online/Targets")
+	for i in targets.get_child_count():
+		targets.get_children()[i].global_transform.origin = pos_array[i]
+
+remote func position_update(pos, rot):
+	if game.get_node("romp_online/dummies") == null:
+		return
+	var actor = game.get_node("romp_online/dummies").get_node(str(get_tree().get_rpc_sender_id()))
+	if actor != null:
+		actor.translation = pos
+		actor.rotation = rot
+
+func start_multiplayer(type):
+	if username_text.text == "":
+		#Abort if no username has been entered.
+		username_text.get_node("warning").visible = true
+		return
+	username = username_text.text
+	
+	#Show waiting screen
+	waiting.visible = true
+	mp_type = type
+	
+	#Create new ENet
+	var peer = NetworkedMultiplayerENet.new()
+	match type:
+		"server":
+			peer.create_server(SERVER_PORT, MAX_PLAYERS)
+			waiting_text.text = "Waiting for players..."
+		"client":
+			peer.create_client(SERVER_IP, SERVER_PORT)
+			waiting_text.text = "Connecting..."
+	get_tree().network_peer = peer
+	
+	#Connect ENet signals
+	get_tree().connect("network_peer_connected", self, "_player_connected")
+	get_tree().connect("network_peer_disconnected", self, "_player_disconnected")
+	get_tree().connect("connected_to_server", self, "_connected_ok")
+	get_tree().connect("connection_failed", self, "_connected_fail")
+	get_tree().connect("server_disconnected", self, "_server_disconnected")
+
+func _player_connected(id):
+	rpc_id(id, "register_player", username)
+	if mp_type == "server":
+		#A player has connected. Start game.
+		start_game()
+
+func _player_disconnected(id):
+	player_info.erase(id) # Erase player from info.
+	if game_running:
+		#Delete dummy
+		game.get_node("romp_online/dummies").get_node(str(id)).queue_free()
+	#Stop game
+	get_tree().network_peer = null
+	get_tree().change_scene("res://menus/playmenu.tscn")
+
+remote func register_player(username):
+	print("Player " + username + " connected!")
+	var id = get_tree().get_rpc_sender_id()
+	#Save username in dictionary
+	player_info[id] = username
```

Comment:
To avoid shadowing the earlier `username` variable I like to make it clear when it's private a la Python. Maybe this is self-evident by it being remote, I still just prefer it be explicit.

```suggestion
remote func register_player(_username):
	print("Player " + _username + " connected!")
	var id = get_tree().get_rpc_sender_id()
	#Save username in dictionary
	player_info[id] = _username
```